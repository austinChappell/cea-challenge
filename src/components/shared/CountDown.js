import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import HelperMethods from '../../helpers/';

const helperMethods = new HelperMethods();

const { getTimerDisplay } = helperMethods;

const propTypes = {
  eventTime: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
  message: PropTypes.string,
  startTime: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
};

const defaultProps = {
  eventTime: null,
  message: null,
  startTime: null,
};

class CountDown extends Component {
  state = {
    time: {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
    },
  }

  componentDidMount() {
    this.startTimer();
  }

  componentWillUnmount() {
    clearInterval(this.stopTimer);
  }

  getTime = () => {
    const { eventTime, startTime } = this.props;
    const now = startTime || Date.now();
    const time = getTimerDisplay(now, eventTime);
    this.setState({ time });
  }

  startTimer = () => {
    this.getTime();
    this.stopTimer = setInterval(() => {
      this.getTime();
    }, 1000);
  }

  render() {
    const {
      days,
      hours,
      minutes,
      seconds,
    } = this.state.time;
    const { message } = this.props;

    return (
      <Fragment>
        <div className="CountDown">
          <div className="time-block">
            <h4>{Number.isNaN(days) ? '--' : days}</h4>
            <span>Days</span>
          </div>
          <div className="time-block">
            <h4>{Number.isNaN(hours) ? '--' : hours}</h4>
            <span>Hours</span>
          </div>
          <div className="time-block">
            <h4>{Number.isNaN(minutes) ? '--' : minutes}</h4>
            <span>Minutes</span>
          </div>
          <div className="time-block">
            <h4>{Number.isNaN(seconds) ? '--' : seconds}</h4>
            <span>Seconds</span>
          </div>
        </div>
        <h6>{message}</h6>
      </Fragment>
    );
  }
}

CountDown.propTypes = propTypes;
CountDown.defaultProps = defaultProps;

export default CountDown;
