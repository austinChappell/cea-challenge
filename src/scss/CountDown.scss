.CountDown {
  display: flex;
  justify-content: center;
  .time-block {
    background-color: $color-black;
    margin: 10px 5px;
    border: 1px solid $color-white;
    border-radius: $standard-border-radius;
    box-sizing: border-box;
    padding: 10px;
    * {
      margin: 0;
    }
  }
}