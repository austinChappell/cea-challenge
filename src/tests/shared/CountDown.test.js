import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import CountDown from '../../components/shared/CountDown';

const props = {
  eventTime: 1522030848733,
  message: '',
};

describe('The CountDown Component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<CountDown {...props} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
