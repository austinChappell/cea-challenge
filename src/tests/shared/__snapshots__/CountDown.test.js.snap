// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CountDown Component should render correctly 1`] = `
<React.Fragment>
  <div
    className="CountDown"
  >
    <div
      className="time-block"
    >
      <h4>
        3449
      </h4>
      <span>
        Days
      </span>
    </div>
    <div
      className="time-block"
    >
      <h4>
        0
      </h4>
      <span>
        Hours
      </span>
    </div>
    <div
      className="time-block"
    >
      <h4>
        23
      </h4>
      <span>
        Minutes
      </span>
    </div>
    <div
      className="time-block"
    >
      <h4>
        20
      </h4>
      <span>
        Seconds
      </span>
    </div>
  </div>
  <h6 />
</React.Fragment>
`;
